src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -628,6 +628,49 @@ public bool Remove(TKey key)
             return false;
         }
 
+        public bool Remove(TKey key, out TValue value)
+        {
+            if (key == null)
+            {
+                value = default(TValue);
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
+            }
+
+            if (buckets != null)
```

Comment:
C# compiler does very few trivial optimizations only. Optimizations like CSE are done by the JIT. Whether or not the optimization kicks in depends on number of factors. The best way to know for sure is to look the disassembly of the generated code and measure.

One thing to keep in mind is that the array indexing in C# is bound-checked. It restricts the optimizations that the JIT can do transparently, for example it cannot do the bound-check speculatively because of it would result in spurious exceptions.

I have glanced over the native code generated for `Dictionary<int,int>.Remove`. The CSE for `entry[i]` kicks in some cases, but not others. If you cache `ref Entry entry = ref entries[i];` at the start of the loop and replace all 9 occurrences of `entries[i]` with it, the native code generated for this method should be approximately 20% smaller.